README.md
[![Project Status](http://opensource.box.com/badges/active.svg)](http://opensource.box.com/badges)
[![Build status](https://ci.appveyor.com/api/projects/status/kbuwacqvl72x6hw6/branch/master?svg=true)](https://ci.appveyor.com/project/coolboy/box-windows-sdk-v2/branch/master)

Box Windows V2 SDK
==================

Windows .NET SDK for V2 of the Box API that is usable from the following frameworks: 
* .NET Framework 4.5
* .NET Core 1.0 or above

Prerequisites
* Visual Studio 2017
* .NET Core SDK (if running .NET Core samples)

Quick Start
-----------

### Installation

Install the SDK using Nuget
```bash
PM> Install-Package Box.V2
```

If you want to use .NET Core
```bash
PM> Install-Package Box.V2.Core
```

If you haven't already created an app in Box go to https://developer.box.com/ and click 'Sign Up'

### Authentication

#### Using a Developer Token (generate one in your app admin console; they last for 60 minutes)
```c#
var config = new BoxConfig(<Client_Id>, <Client_Secret>, new Uri("http://localhost"));
var session = new OAuthSession(<Developer_Token>, "NOT_NEEDED", 3600, "bearer");
client = new BoxClient(config, session);
```

#### Using with Box Platform Developer or Box Platform Enterprise

#### When to Use Box Platform (Enterprise and Developer)
*Box Platform Enterprise*
This represents your application within a Box enterprise. Use this type of authentication if you are trying to:
Store content at the application level rather than at the individual user level. 
Access any user and make as-user api calls to impersonate a user. 
Apply Enterprise features such as: managing and applying retention policies, using metadata object, and accessing Events endpoint of the content api.

*Box Platform Developer*
Box account that belongs to your Box Platform application, this is different from an end-user of Box. These accounts do not have an associated login and 
can only be accessed through the Box API. Use this type of authentication if you are trying to:
Have Box content management functionalities in an external-facing app - customer portal
Provide access to content stored in Box to internal users who do not have Box Managed User accounts. 
Allow Box Managed Users to share and collaboration with external users via Box applications

##### Configure
```c#
var boxConfig = new BoxConfig(<Client_Id>, <Client_Secret>, <Enterprise_Id>, <Private_Key>, <JWT_Private_Key_Password>, <JWT_Public_Key_Id>);
var boxJWT = new BoxJWTAuth(boxConfig);
```

##### Authenticate
```c#
var adminToken = boxJWT.AdminToken(); //valid for 60 minutes so should be cached and re-used
var adminClient = boxJWT.AdminClient(adminToken);
```

##### Create an App User
```c#
//NOTE: you must set IsPlatformAccessOnly=true for an App User
var userRequest = new BoxUserRequest() { Name = "test appuser", IsPlatformAccessOnly = true };
var appUser = await adminClient.UsersManager.CreateEnterpriseUserAsync(userRequest);

//get a user client
var userToken = boxJWT.UserToken(appUser.Id); //valid for 60 minutes so should be cached and re-used
var userClient = boxJWT.UserClient(userToken, appUser.Id);

//for example, look up the app user's details
var userDetails = await userClient.UsersManager.GetCurrentUserInformationAsync();
```

#### Using with OAuth2

#### When to Use Box Integration (Oauth 2.0)
Oauth 2.0 requires user to log in to Box and grant your application permission to access files and folders.
This is a three-legged authentication process to allow managed user and external users to interact.

##### Configure
Set your configuration parameters and initialize the client:
```c#
var config = new BoxConfig(<Client_Id>, <Client_Secret>, <Redirect_Uri>);
var client = new BoxClient(config);
```

##### Authenticate
Bundled with the SDK are sample applications for both Windows 8 and Windows Phone which include sample OAuth2 Workflows. The authentication workflow is a 2-step process that first retrieves an Auth Code and then exchanges it for an Access/Refresh Token

*Windows 8*
```c#
string authCode = await OAuth2Sample.GetAuthCode(config.AuthCodeUri, new Uri(config.RedirectUri));
await client.Auth.AuthenticateAsync(authCode);
```

*Windows Phone*
```c#
// Ensure the OAuth2Sample control is placed at the root level of the application page xaml and named "oAuth2Sample"
// Subscribe to the received call back 
oAuth2Sample.AuthCodeReceived += async (s, e) =>
{
    var auth = s as OAuth2Sample;
    await client.Auth.AuthenticateAsync(auth.AuthCode);
};
// Navigate and show the login page
oauth.GetAuthCode(config.AuthCodeUri, config.RedirectUri);
```

*Other (ASP.NET)*

Alternatively, a completely custom OAuth2 authentication process can be used in place of the provided workflows, for example, in a custom web application. In this scenario, a fully formed OAuthSession object should be passed in when instantiating the BoxClient. 

```c#
OAuthSession session = // Create session from custom implementation
var client = new BoxClient(config, session);
```

#### Using With App Token Auth (Box View)

Box View uses App Token Auth, where the app generated long-lived access tokens
that are used directly in place of any authentication calls to the API.  To use
this method of aythentication, simply create a client with only the API key and
access token provided:

```c#
var config = new BoxConfig(<API_KEY>, "", new Uri("http://localhost"));
var session = new OAuthSession(<PRIMARY OR SECONDARY TOKEN>, "NOT_NEEDED", 3600, "bearer");
client = new BoxClient(config, session);
```

### Examples
#### Get Folder Items
```c#
// Get root folder with default properties
var items = await client.FoldersManager.GetFolderItemsAsync("0", 500);
```

#### Get File Information
```c#
BoxFile f = await client.FilesManager.GetInformationAsync(fileId);
```

#### Update a Files Information
```c#
// Create request object with new property values
BoxFileRequest request = new BoxFileRequest()
{
    Id = fileId,
    Name = "NewName",
    Description = "New Description"
};
BoxFile f = await client.FilesManager.UpdateInformationAsync(request );
```

#### Upload a New File
```c#
BoxFile newFile;

// Create request object with name and parent folder the file should be uploaded to
using (FileStream stream = new FileStream(@"C:\\example.pdf", FileMode.Open))
{
	BoxFileRequest req = new BoxFileRequest()
	{
		Name = "example.pdf",
		Parent = new BoxRequestEntity() { Id = "0" }
	};
	newFile = await client.FilesManager.UploadAsync(req, stream);
}
```

#### Upload a New File with Content MD5 hash
```c#
BoxFile newFile;

// Create request object with name and parent folder the file should be uploaded to
using (FileStream stream = new FileStream(@"C:\\example.pdf", FileMode.Open))
using (SHA1 sha1 = SHA1.Create())
{
	BoxFileRequest req = new BoxFileRequest()
	{
		Name = "example.pdf",
		Parent = new BoxRequestEntity() { Id = "0" }
	};
	
	byte[] md5Bytes = sha1.ComputeHash(fs);
	
	newFile = await client.FilesManager.UploadAsync(req, stream, contentMD5: md5Bytes);
}
```

#### Perform Preflight Check for a new file upload
```c#
try
{
	var req = new BoxPreflightCheckRequest() { 
	    Name = "example.pdf", 
	    Parent = new BoxRequestEntity() { Id = "0" },
	    Size = 10000 //set the size if known, otherwise don't set (i.e. for a stream)
	};
											 
	//exception will be thrown if name collision or storage limit would be exceeded by upload									 
	await userClient.FilesManager.PreflightCheck(req);
}
catch (BoxPreflightCheckConflictException<BoxFile> bex)
{
	//Handle file name collision error	
}
catch (BoxException bex)
{
	//Handle storage limit error 
}
```

#### Perform Preflight Check for a new version of file
```c#
try
{
	var req = new BoxPreflightCheckRequest() { Size=10926 };
	
	//exception will be thrown if storage limit would be exceeded by uploading new version of file
        await userClient.FilesManager.PreflightCheckNewVersion(existingFile.Id, req);		 
}
catch (BoxException bex)
{
	//Handle storage limit error
}
```

#### Download a File
```c#
Stream stream = await client.FilesManager.DownloadStreamAsync(fileId);
```

#### Get Temporary Download Uri for a file
This method will retrieve a temporary (15 minute) Uri for a file that can be used, for example, to send as a redirect to a browser, causing the browser to download the file directly from Box.
```c#
var downloadUri = await client.FilesManager.GetDownloadUriAsync(fileId);
```

#### Search using Metadata
```c#
var filter = new 
{ 
	someKey = "blah", 
	expiresOn = new {gt = new DateTime(2015,1,1), 
			 lt = new DateTime(2015,9,1)},
	count = new {gt = 5, lt = 10},
	option = "value1"
};

var mdFilter = new BoxMetadataFilterRequest() 
{ 
	TemplateKey = "yourTemplate", 
	Scope = "enterprise", 
	Filters = filter 
};

//currently only one BoxMetadataFilterRequest element is supported; in the future multiple will be supported (hence the List)
var results = await client.SearchManager.SearchAsync(mdFilters: new List<BoxMetadataFilterRequest>() { mdFilter });
```

#### Make API calls with As-User

#### Example 1
If you have an admin token with appropriate permissions, you can make API calls in the context of a managed user. In order to do this you must request Box.com to activate As-User functionality for your API key (see developer site for instructions). 

```c#
var config = new BoxConfig(<Client_Id>, <Client_Secret>, <Redirect_Uri);
var auth = new OAuthSession(<Your_Access_Token>, <Your_Refresh_Token>, 3600, "bearer");

var userId = "12345678"
var userClient = new BoxClient(config, auth, asUser: userId);

//returns root folder items for the user with ID '12345678'
var items  = await userClient.FoldersManager.GetFolderItemsAsync("0", 500);
```

#### Example 2

Using the admin token we can make a call to retrieve all users or a specific user

```cs
var boxConfig = new BoxConfig(<Client_Id>, <Client_Secret>, <Enterprise_Id>, <Private_Key>, <JWT_Private_Key_Password>, <JWT_Public_Key_Id>);
var boxJWT = new BoxJWTAuth(boxConfig);

var adminToken = boxJWT.AdminToken();
var adminClient = boxJWT.AdminClient(adminToken);

var boxUsers = await adminClient.UsersManager.GetEnterpriseUsersAsync();
List<BoxUser> allBoxUsersList = boxUsers.Entries;

// Display all users with name and id per row
foreach(BoxUser boxUser in allBoxUsersList)

CHANGELOG.md
# Changelog

## 3.24.0 [2020-07-21]
- Add path parameter sanitization
- Add support for the classification field for Files and Folders
- Fix bug with notification email field deserializing for `BoxUser` 
- Add `fields` parameter for metadata query
- Add ability to set a request timeout for `FoldersManager.UpdateInformationAsync()` and `UsersManager.MoverUserFolderAsync`

## 3.23.0 [2020-05-12]
- Add ability to get and set a notification email address for a user
- Fix deadlock issue for JWT authentication in UI elements
- Add support for the uploader display name field for Files and File Versions

## 3.22.0 [2020-02-25]
- Fixed Authentication Request Retries
- Added the ability to query Box items based on their metadata. The method to do so is `MetadataManager.ExecuteMetadataQueryAsync()`.
- Added `TrashedAt` field to `BoxItem` objects (file, folder, weblink).
- Added marker based pagination for get users methods
- Updated retry logic to retry on 503 status codes returned by the API
- Provide better details for debugging, if the HttpClient used to make API requests times out

## 3.21.0 [2019-12-05]
- Added `fields` parameter to `UsersManager.GetUserInformationAsync()`
- Added `ExternalAppUserId` property to `BoxUser` model
- Added the ability to set the `TrackingCodes` property when updating or creating a user (thanks @Cpcrook!)

## 3.20.0 [2019-09-19]
- Added missing fields for File Version object.

## 3.19.0 [2019-08-29]
- Added `FILE_VERSION_RESTORE` constant to Admin Event.
- Added action_by field to Enterprise Event.
- Audited missing fields on BoxFile and BoxFolder objects.
- Better error handling and messaging for errors pertaining to OAuth2 error responses.

## 3.18.0 [2019-06-20]

- Added `sort` and `direction` parameters to `FoldersManager.GetFolderItemsAsync()` to enable
  sorting the folder items returned
- Added a new `SearchManager.QueryAsync()` method with correct types for file size filter parameters
- Deprecated the `SearchManager.SearchAsync()` method, which is superseded by `SearchManager.QueryAsync()`
- Added support for setting the `IsExternalCollabRestricted` parameter when creating and updating Users
- Added a `WebProxy` property to `BoxConfig` instances, which can be used to manually set the network proxy
  used by the SDK

## 3.17.0 [2019-05-09]

- Fixed the encoding of dates in the query parameters for Events and Search endpoints
- Deprecated `FilesManager.DownloadStreamAsync()` and introduced a replacement method with correct parameter types for byte offsets: `FilesManager.DownloadAsync()`

## 3.16.0 [2019-04-29]

- Added `sort` and `direction` parameters to `client.SearchManager.SearchAsync()` to  control sort order
- Added `extension` parameter to `client.FilesManager.GetThumbnailAsync()` to control which thumbnail format is returned (thanks @guilmori!)
- Fixed a bug where query string parameters were not correctly encoded
- Added `SetFileMetadataAsync()` and `SetFolderMetadataAsync()` methods to `client.MetadataManager` to set metadata
  keys and values, overwriting existing values for the provided keys.
- Automatically retry most API calls when the API responds with a transient error status code

## 3.15.0 [2019-03-28]

- Added support for passing custom IBoxService to BoxJWTAuth constructor.

## 3.14.1 [2019-03-07]

- Removed unnecessary package.config from sample files.

## 3.14.0 [2019-02-28]

- Added trace ID to API response exception message.
- Fix deserialization of translated task assignment status.

## 3.13.1 [2019-02-21]

- Fixed an issue where some objects related to Events did not have their `.Id` property correctly deserialized from JSON

## 3.13.0 [2019-02-14]

- Added the `.InviteEmail` property to `BoxCollaboration` objects, which displays the email address for the invited
  user in a pending collaboration
- Added `.Timezone`, `.IsExternalCollabRestricted`, `.Tags`, and `.Hostname` properties to `BoxUser` objects

## 3.12.0 [2019-02-07]

- Added `client.FilesManager.GetCollaborationsCollectionAsync()` and deprecated
  `client.FilesManager.GetCollaborationsAsync()` to enable paging through the entire
  collection of collaborations on a file
- Added `client.WebLinksManager.CopyAsync()`, `client.WebLinksManager.CreateSharedLinkAsync()`,
  and `client.WebLinksManager.DeleteSharedLinkAsync()`
- Added `client.UsersManager.GetUserAvatarAsync()` for retrieving a user's avatar image

## 3.11.0 [2019-01-17]

- Added support for reading and writing more Group fields
- Fixed an issue where the `UnsharedAt` field of a shared link could not be set to `null`
- Fixed renaming a file on new version upload
- Added the ability to set the content modification timestamp on file version upload
- Fixed issues around reading the source of an event when the source item is a web link

## 3.10.0 [2018-12-14]

- Added support for Metadata Cascade Policies

## 3.9.3 [2018-09-04]

- Strong named the assembly.

## 3.9.2 [2018-06-14]

- Added support for [setting flag](https://github.com/box/box-windows-sdk-v2/blob/master/Box.V2/Models/Request/BoxFolderRequest.cs#L39) allowing non owners of a folder to invite collaborators.

## 3.9.1 [2018-06-07]


Box.V2.Test/BoxGroupsManagerTest.cs
using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Box.V2.Managers;
using System.Threading.Tasks;
using Box.V2.Models;
using Moq;
using Box.V2.Models.Request;

namespace Box.V2.Test
{
    [TestClass]
    public class BoxGroupsManagerTest : BoxResourceManagerTest
    {
        private readonly BoxGroupsManager _groupsManager;

        public BoxGroupsManagerTest()
        {
            _groupsManager = new BoxGroupsManager(Config.Object, Service, Converter, AuthRepository);
        }

        [TestMethod]
        [TestCategory("CI-UNIT-TEST")]
        public async Task GetGroupItems_ValidResponse_ValidGroups()
        {
            Handler.Setup(h => h.ExecuteAsync<BoxCollection<BoxGroup>>(It.IsAny<IBoxRequest>()))
                .Returns(() => Task.FromResult<IBoxResponse<BoxCollection<BoxGroup>>>(new BoxResponse<BoxCollection<BoxGroup>>()
                {
                    Status = ResponseStatus.Success,
                    ContentString = @"{""total_count"": 14, 
                                    ""entries"": [  {""type"": ""group"", ""id"": ""26477"", ""name"": ""adfasdf"", ""created_at"": ""2011-02-15T14:07:22-08:00"", ""modified_at"": ""2011-10-05T19:04:40-07:00""}, 
                                                    {""type"": ""group"", ""id"": ""1263"", ""name"": ""Enterprise Migration"", ""created_at"": ""2009-04-20T19:36:17-07:00"", ""modified_at"": ""2011-10-05T19:05:10-07:00""}],
                                    ""limit"": 2, ""offset"": 0}"
                }));

            BoxCollection<BoxGroup> items = await _groupsManager.GetAllGroupsAsync();

            Assert.AreEqual(items.TotalCount, 14, "Wrong total count");
            Assert.AreEqual(items.Entries.Count, 2, "Wrong count of entries");
            Assert.AreEqual(items.Entries[0].Type, "group", "Wrong type");
            Assert.AreEqual(items.Entries[0].Id, "26477", "Wrong id");
            Assert.AreEqual(items.Entries[0].Name, "adfasdf", "Wrong name");
            Assert.AreEqual(items.Entries[0].CreatedAt, DateTime.Parse("2011-02-15T14:07:22-08:00"), "Wrong created at");
            Assert.AreEqual(items.Entries[0].ModifiedAt, DateTime.Parse("2011-10-05T19:04:40-07:00"), "Wrong modified at");
            Assert.AreEqual(items.Offset, 0, "Wrong offset");
            Assert.AreEqual(items.Limit, 2, "Wrong limit");
        }

        [TestMethod]
        [TestCategory("CI-UNIT-TEST")]
        public async Task GetGroup_ValidResponse_ValidGroup()
        {
            Handler.Setup(h => h.ExecuteAsync<BoxGroup>(It.IsAny<IBoxRequest>()))
                .Returns(() => Task.FromResult<IBoxResponse<BoxGroup>>(new BoxResponse<BoxGroup>()
                {
                    Status = ResponseStatus.Success,
                    ContentString = @"{""type"": ""group"", ""id"": ""26477"", ""name"": ""adfasdf"", ""created_at"": ""2011-02-15T14:07:22-08:00"", ""modified_at"": ""2011-10-05T19:04:40-07:00""}"
                }));

            BoxGroup group = await _groupsManager.GetGroupAsync("222");

            Assert.AreEqual("26477", group.Id, "Wrong Id");
            Assert.AreEqual("group", group.Type, "Wrong type");
            Assert.AreEqual("adfasdf", group.Name, "Wrong name");
            Assert.AreEqual(DateTime.Parse("2011-02-15T14:07:22-08:00"), group.CreatedAt, "Wrong created at");
            Assert.AreEqual(DateTime.Parse("2011-10-05T19:04:40-07:00"), group.ModifiedAt, "Wrong modified at");
        }

        [TestMethod]
        [TestCategory("CI-UNIT-TEST")]
        public async Task GetGroupItems_ValidResponse_NoGroups()
        {
            Handler.Setup(h => h.ExecuteAsync<BoxCollection<BoxGroup>>(It.IsAny<IBoxRequest>()))
                .Returns(() => Task.FromResult<IBoxResponse<BoxCollection<BoxGroup>>>(new BoxResponse<BoxCollection<BoxGroup>>()
                {
                    Status = ResponseStatus.Success,
                    ContentString = @"{""total_count"": 0, 
                                    ""limit"": 2, ""offset"": 0}"
                }));

            BoxCollection<BoxGroup> items = await _groupsManager.GetAllGroupsAsync();

            Assert.AreEqual(0, items.TotalCount, "Wrong count");
            Assert.AreEqual(0, items.Offset, "Wrong offset");
            Assert.AreEqual(2, items.Limit, "Wrong limit");
        }

        [TestMethod]
        [TestCategory("CI-UNIT-TEST")]
        public async Task CreateGroup_ValidResponse_NewGroup()
        {
            Handler.Setup(h => h.ExecuteAsync<BoxGroup>(It.IsAny<BoxRequest>()))
                .Returns(() => Task.FromResult<IBoxResponse<BoxGroup>>(new BoxResponse<BoxGroup>()
                {
                    Status = ResponseStatus.Success,
                    ContentString = @"{ ""type"": ""group"", ""id"": ""159322"", ""name"": ""TestGroup2"", ""created_at"": ""2013-11-12T15:19:47-08:00"", ""modified_at"": ""2013-11-12T15:19:47-08:00"" }"
                }));

            BoxGroupRequest request = new BoxGroupRequest(){ Name = "NewGroup" };

            BoxGroup group = await _groupsManager.CreateAsync(request);

            Assert.AreEqual<string>("TestGroup2", group.Name, "Wrong group name");
            Assert.AreEqual<string>("159322", group.Id, "Wrong id");
            Assert.AreEqual(DateTime.Parse("2013-11-12T15:19:47-08:00"), group.ModifiedAt, "Wrong modified at");
            Assert.AreEqual(DateTime.Parse("2013-11-12T15:19:47-08:00"), group.CreatedAt, "Wrong created at");
        }

        [TestMethod]
        [TestCategory("CI-UNIT-TEST")]
        public async Task DeleteGroup_ValidResponse_ValidGroup()
        {
            Handler.Setup(h=>h.ExecuteAsync<BoxGroup>(It.IsAny<IBoxRequest>()))
                .Returns(() => Task.FromResult<IBoxResponse<BoxGroup>>(new BoxResponse<BoxGroup>()
                {
                    Status = ResponseStatus.Success,
                    ContentString = ""
                }));

            var result = await _groupsManager.DeleteAsync("1234");
            Assert.IsTrue(result, "Unsuccessful delete");
        }

        [TestMethod]
        [TestCategory("CI-UNIT-TEST")]
        public async Task UpdateGroup_ValidResponse_ValidGroup()
        {
            Handler.Setup(h => h.ExecuteAsync<BoxGroup>(It.IsAny<IBoxRequest>()))
                .Returns(() => Task.FromResult<IBoxResponse<BoxGroup>>(new BoxResponse<BoxGroup>()
                {
                    Status = ResponseStatus.Success,
                    ContentString = @"{ ""type"": ""group"", ""id"": ""159322"", ""name"": ""TestGroup2"", ""created_at"": ""2013-11-12T15:19:47-08:00"", ""modified_at"": ""2013-11-12T15:19:47-08:00"" }"
                }));

            BoxGroupRequest request = new BoxGroupRequest() { Name = "groupName" };

            BoxGroup group = await _groupsManager.UpdateAsync("123", request);

            Assert.AreEqual<string>("TestGroup2", group.Name, "Wrong Group name");
            Assert.AreEqual<string>("159322", group.Id, "Wrong group id");
            Assert.AreEqual(DateTime.Parse("2013-11-12T15:19:47-08:00"), group.ModifiedAt, "Wrong modified at");
            Assert.AreEqual(DateTime.Parse("2013-11-12T15:19:47-08:00"), group.CreatedAt, "Wrong created at");
        }

        [TestMethod]
        [TestCategory("CI-UNIT-TEST")]
        public async Task UpdateGroup_ExtraFields_ValidGroup()
        {
            IBoxRequest boxRequest = null;
            Handler.Setup(h => h.ExecuteAsync<BoxGroup>(It.IsAny<IBoxRequest>()))
                .Returns(() => Task.FromResult<IBoxResponse<BoxGroup>>(new BoxResponse<BoxGroup>()
                {
                    Status = ResponseStatus.Success,
                    ContentString = @"{
                        ""type"": ""group"",
                        ""id"": ""159322"",
                        ""description"": ""A group from Okta"",
                        ""external_sync_identifier"": ""foo"",
                        ""provenance"": ""Okta"",
                        ""invitability_level"": ""admins_only"",
                        ""member_viewability_level"": ""admins_only""
                    }"
                }))
                .Callback<IBoxRequest>(r => boxRequest = r);

            BoxGroupRequest request = new BoxGroupRequest() {
                Description = "A group from Okta",
                ExternalSyncIdentifier = "foo",
                Provenance = "Okta",
                InvitabilityLevel = "admins_only",
                MemberViewabilityLevel = "admins_only"
            };

            var fields = new string[]
            {
               BoxGroup.FieldDescription,
               BoxGroup.FieldExternalSyncIdentifier,
               BoxGroup.FieldProvenance,
               BoxGroup.FieldInvitabilityLevel,
               BoxGroup.FieldMemberViewabilityLevel
            };

            BoxGroup group = await _groupsManager.UpdateAsync("123", request, fields: fields);

            Assert.AreEqual("{\"description\":\"A group from Okta\",\"provenance\":\"Okta\",\"external_sync_identifier\":\"foo\",\"invitability_level\":\"admins_only\",\"member_viewability_level\":\"admins_only\"}", boxRequest.Payload);
            Assert.AreEqual("A group from Okta", group.Description);
            Assert.AreEqual("foo", group.ExternalSyncIdentifier);
            Assert.AreEqual("Okta", group.Provenance);
            Assert.AreEqual("admins_only", group.InvitabilityLevel);
            Assert.AreEqual("admins_only", group.MemberViewabilityLevel);
        }

        [TestMethod]
        [TestCategory("CI-UNIT-TEST")]
        public async Task AddGroupMembership_ValidResponse()
        {
            Handler.Setup(h => h.ExecuteAsync<BoxGroupMembership>(It.IsAny<IBoxRequest>()))
                .Returns(() => Task.FromResult<IBoxResponse<BoxGroupMembership>>(new BoxResponse<BoxGroupMembership>()
                {
                    Status = ResponseStatus.Success,
                    ContentString = @"{ ""type"": ""group_membership"", ""id"": ""2381570"",
                                        ""user"": { ""type"": ""user"", ""id"": ""4940223"", ""name"": ""<script>alert(12);</script>"", ""login"": ""testmaster@box.net""},
                                        ""group"": { ""type"": ""group"", ""id"": ""159622"", ""name"": ""test204bbee0-f3b0-43b4-b213-214fcd2cb9a7"" },
                                        ""role"": ""member"",
                                        ""created_at"": ""2013-11-13T13:19:44-08:00"",
                                        ""modified_at"": ""2013-11-13T13:19:44-08:00""}"
                }));

            BoxGroupMembershipRequest request = new BoxGroupMembershipRequest()
            {
                User = new BoxRequestEntity() { Id = "123" },
                Group = new BoxGroupRequest() { Id = "456" }
            };

            BoxGroupMembership response = await _groupsManager.AddMemberToGroupAsync(request);

            Assert.AreEqual("2381570", response.Id, "Wrong Membership id");
            Assert.AreEqual("group_membership", response.Type, "wrong type");
            Assert.AreEqual("159622", response.Group.Id, "Wrong group id");
            Assert.AreEqual("4940223", response.User.Id, "Wrong user id");
            Assert.AreEqual("member", response.Role, "Wrong role");
            Assert.AreEqual(DateTime.Parse("2013-11-13T13:19:44-08:00"), response.CreatedAt, "Wrong created at");
            Assert.AreEqual(DateTime.Parse("2013-11-13T13:19:44-08:00"), response.ModifiedAt, "Wrong modified at");
        }

        [TestMethod]
        [TestCategory("CI-UNIT-TEST")]
        public async Task DeleteMembership_ValidResponse_ValidGroup()
        {
            Handler.Setup(h => h.ExecuteAsync<BoxGroup>(It.IsAny<IBoxRequest>()))
                .Returns(() => Task.FromResult<IBoxResponse<BoxGroup>>(new BoxResponse<BoxGroup>()
                {
                    Status = ResponseStatus.Success,
                    ContentString = ""
                }));

            var result = await _groupsManager.DeleteGroupMembershipAsync("1234");
            Assert.IsTrue(result, "Unsuccessful group membership delete");
        }
        
        [TestMethod]
        [TestCategory("CI-UNIT-TEST")]
        public async Task GetAllMemberships_ValidResponse_ValidGroup()
        {
            Handler.Setup(h => h.ExecuteAsync<BoxCollection<BoxGroupMembership>>(It.IsAny<IBoxRequest>()))
                .Returns(() => Task.FromResult<IBoxResponse<BoxCollection<BoxGroupMembership>>>(new BoxResponse<BoxCollection<BoxGroupMembership>>()
                {
                    Status = ResponseStatus.Success,
                    ContentString = @"{ ""total_count"": 2, ""entries"": [{
                                        ""type"": ""group_membership"", ""id"": ""136639"",
                                        ""user"": {""type"": ""user"",""id"": ""6102564"", ""name"": ""<script>alert('atk');</script>"", ""login"": ""testanyregister@box.net""},
                                        ""group"": {""type"": ""group"",""id"": ""26477"",""name"": ""adfasdf""},""role"": ""member""}, 
                                        {""type"": ""group_membership"",""id"": ""273529"",
                                        ""user"": {""type"": ""user"",""id"": ""13928063"",""name"": ""spootie"",""login"": ""tevanspratt++39478@box.net""},
                                        ""group"": {""type"": ""group"", ""id"": ""26477"", ""name"": ""adfasdf""}, ""role"": ""member""}], ""offset"": 0,""limit"": 100}"
                }));
            
            BoxCollection<BoxGroupMembership> response = await _groupsManager.GetAllGroupMembershipsForGroupAsync("123");

            Assert.AreEqual(2, response.TotalCount, "Wrong total count");
            Assert.AreEqual(2, response.Entries.Count, "Wrong number of entries");
            Assert.AreEqual("group_membership", response.Entries[0].Type, "Wrong type");
            Assert.AreEqual("6102564", response.Entries[0].User.Id, "Wrong user id");
            Assert.AreEqual("26477", response.Entries[0].Group.Id, "Wrong group id");
        }

        [TestMethod]
        [TestCategory("CI-UNIT-TEST")]
        public async Task GetAllMemberships_ValidResponse_ValidUser()
        {
            Handler.Setup(h => h.ExecuteAsync<BoxCollection<BoxGroupMembership>>(It.IsAny<IBoxRequest>()))
                .Returns(() => Task.FromResult<IBoxResponse<BoxCollection<BoxGroupMembership>>>(new BoxResponse<BoxCollection<BoxGroupMembership>>()
                {
                    Status = ResponseStatus.Success,
                    ContentString = @"{ ""total_count"": 2, ""entries"": [{
                                        ""type"": ""group_membership"", ""id"": ""136639"",
                                        ""user"": {""type"": ""user"",""id"": ""6102564"", ""name"": ""<script>alert('atk');</script>"", ""login"": ""testanyregister@box.net""},
                                        ""group"": {""type"": ""group"",""id"": ""26477"",""name"": ""adfasdf""},""role"": ""member""}, 
                                        {""type"": ""group_membership"",""id"": ""273529"",
                                        ""user"": {""type"": ""user"",""id"": ""13928063"",""name"": ""spootie"",""login"": ""tevanspratt++39478@box.net""},
                                        ""group"": {""type"": ""group"", ""id"": ""26477"", ""name"": ""adfasdf""}, ""role"": ""member""}], ""offset"": 0,""limit"": 100}"
                }));

            BoxCollection<BoxGroupMembership> response = await _groupsManager.GetAllGroupMembershipsForUserAsync("123");

            Assert.AreEqual(2, response.TotalCount, "Wrong total count");
            Assert.AreEqual(2, response.Entries.Count, "Wrong number of entries");
            Assert.AreEqual("group_membership", response.Entries[0].Type, "Wrong type");
            Assert.AreEqual("6102564", response.Entries[0].User.Id, "Wrong user id");
            Assert.AreEqual("26477", response.Entries[0].Group.Id, "Wrong group id");
        }

        [TestMethod]
        [TestCategory("CI-UNIT-TEST")]
        public async Task GetGroupMembership_ValidResponse()
        {
            Handler.Setup(h => h.ExecuteAsync<BoxGroupMembership>(It.IsAny<IBoxRequest>()))
                .Returns(() => Task.FromResult<IBoxResponse<BoxGroupMembership>>(new BoxResponse<BoxGroupMembership>()
                {
                    Status = ResponseStatus.Success,
                    ContentString = @"{ ""type"": ""group_membership"", ""id"": ""2381570"",
                                        ""user"": { ""type"": ""user"", ""id"": ""4940223"", ""name"": ""<script>alert(12);</script>"", ""login"": ""testmaster@box.net""},
                                        ""group"": { ""type"": ""group"", ""id"": ""159622"", ""name"": ""test204bbee0-f3b0-43b4-b213-214fcd2cb9a7"" },
                                        ""role"": ""member"",
                                        ""created_at"": ""2013-11-13T13:19:44-08:00"",
                                        ""modified_at"": ""2013-11-13T13:19:44-08:00""}"
                }));

            BoxGroupMembership response = await _groupsManager.GetGroupMembershipAsync("123");

            Assert.AreEqual("2381570", response.Id, "Wrong Membership id");
            Assert.AreEqual("group_membership", response.Type, "wrong type");
            Assert.AreEqual("159622", response.Group.Id, "Wrong group id");
            Assert.AreEqual("4940223", response.User.Id, "Wrong user id");
            Assert.AreEqual("member", response.Role, "Wrong role");
            Assert.AreEqual(DateTime.Parse("2013-11-13T13:19:44-08:00"), response.CreatedAt, "Wrong created at");
            Assert.AreEqual(DateTime.Parse("2013-11-13T13:19:44-08:00"), response.ModifiedAt, "Wrong modified at");
        }

        [TestMethod]
        [TestCategory("CI-UNIT-TEST")]
        public async Task UpdateGroupMembership_ValidResponse()
        {
            Handler.Setup(h => h.ExecuteAsync<BoxGroupMembership>(It.IsAny<IBoxRequest>()))
                .Returns(() => Task.FromResult<IBoxResponse<BoxGroupMembership>>(new BoxResponse<BoxGroupMembership>()
                {
                    Status = ResponseStatus.Success,
                    ContentString = @"{ ""type"": ""group_membership"", ""id"": ""2381570"",
                                        ""user"": { ""type"": ""user"", ""id"": ""4940223"", ""name"": ""<script>alert(12);</script>"", ""login"": ""testmaster@box.net""},
                                        ""group"": { ""type"": ""group"", ""id"": ""159622"", ""name"": ""test204bbee0-f3b0-43b4-b213-214fcd2cb9a7"" },
                                        ""role"": ""admin"",
                                        ""created_at"": ""2013-11-13T13:19:44-08:00"",
                                        ""modified_at"": ""2013-11-13T13:19:44-08:00""}"
                }));

            BoxGroupMembershipRequest request = new BoxGroupMembershipRequest() { Role = "anything"};

            BoxGroupMembership response = await _groupsManager.UpdateGroupMembershipAsync("123", request);

            Assert.AreEqual("2381570", response.Id, "Wrong Membership id");
            Assert.AreEqual("group_membership", response.Type, "wrong type");
            Assert.AreEqual("159622", response.Group.Id, "Wrong group id");
            Assert.AreEqual("4940223", response.User.Id, "Wrong user id");
            Assert.AreEqual("admin", response.Role, "Wrong role");
            Assert.AreEqual(DateTime.Parse("2013-11-13T13:19:44-08:00"), response.CreatedAt, "Wrong created at");
            Assert.AreEqual(DateTime.Parse("2013-11-13T13:19:44-08:00"), response.ModifiedAt, "Wrong modified at");
        }
    }
}
